sdk/core/System.ClientModel/src/Message/BinaryContent.cs

```diff
@@ -191,4 +200,45 @@ public override void Dispose()
             }
         }
     }
+
+    private sealed class StreamBinaryContent : BinaryContent
+    {
+        private readonly Stream _stream;
+        private readonly long _origin;
+
+        public StreamBinaryContent(Stream stream)
+        {
+            if (!stream.CanSeek)
+            {
+                throw new ArgumentException("Stream must be seekable.", nameof(stream));
+            }
+
+            _stream = stream;
+            _origin = stream.Position;
+        }
+
+        public override bool TryComputeLength(out long length)
+        {
+            // CanSeek is validated in constructor - it will always be true.
+            length = _stream.Length - _origin;
+            return true;
+        }
+
+        public override void WriteTo(Stream stream, CancellationToken cancellationToken)
+        {
+            _stream.Seek(_origin, SeekOrigin.Begin);
+            _stream.CopyTo(stream, cancellationToken);
+        }
+
+        public override async Task WriteToAsync(Stream stream, CancellationToken cancellation)
+        {
+            _stream.Seek(_origin, SeekOrigin.Begin);
+            await _stream.CopyToAsync(stream, cancellation).ConfigureAwait(false);
```

Comment:
should we lush at the end?

Reply:
Remind me the argument in favor of doing this?

Reply:
somebody needs to flush the writes. If no code flushes, the bytes might not be sent over the wire. So a good practice if: if you are not sure whether there will be flush after you completed all your writes, you should flush

Reply:
Discussed offline - there is a perf cost to adding the flush, but since we've already written so much, we can afford this.